README.md
log.py
======

Simple, clever, pythonic logging.


CHANGES.rst
0.1.3 (2013-10-3)
    Logger.add_output() now passes argument to Output() if it is not
        an Output instance.

    Added todo list.

0.1.2 (2013-10-3)
    Remove log.py

0.1.1 (2013-10-2)
    Add /tests directory
    Add tests
    Add compatibility module: five.py

0.1.0 (2013-07-15)
    Initial Commit


tests/test_five.py
import unittest          # Unittest
import sys               # For version purposes

from log import five     # Module under test


def version_skip(version):
    message = "test only applies to Python {0}".format(version)
    return unittest.skipUnless(sys.version_info[0] == version, message)


class TestFiveShim(unittest.TestCase):

    @version_skip(2)
    def test_string_classes_py2(self):
        self.assertEqual(five.str, basestring)

    @version_skip(3)
    def test_string_classes_py3(self):
        self.assertEqual(five.str, str)

    @version_skip(2)
    def test_is_py2(self):
        self.assertTrue(five.PY2)

    @version_skip(3)
    def test_is_py3(self):
        self.assertTrue(five.PY3)

    @version_skip(2)
    def test_stringio_py2(self):
        self.assertEqual(five.StringIO, five.BytesIO)

    @version_skip(3)
    def test_stringio_py3(self):
        self.assertEqual(five.StringIO, five.cStringIO)


tests/test_levels.py
import unittest          # Unittest

import log               # Module under test


class TestLevelFunction(unittest.TestCase):
    """
    Testing the log.level() function.
    """

    def test_returnsfunc(self):
        """
        Ensure that log.level() returns a function.
        """
        self.assertTrue(callable(log.level(4)))
        self.assertTrue(callable(log.level("INFO")))

    def test_filter(self):
        """
        Test that the filter returned by log.level works on
        strings and numbers.
        """
        filtr = log.level("INFO")

        self.assertTrue(filtr("INFO"))
        self.assertFalse(filtr("DEBUG"))
        self.assertTrue(filtr("ERROR"))

        self.assertTrue(filtr(20))
        self.assertFalse(filtr(4))
        self.assertTrue(filtr(100))

        filtr2 = log.level(25)
        self.assertTrue(filtr2("NOTICE"))
        self.assertFalse(filtr2("DEBUG"))
        self.assertTrue(filtr2("ERROR"))

        self.assertTrue(filtr2(25))
        self.assertFalse(filtr2(4))
        self.assertTrue(filtr2(100))


tests/test_logger.py
import unittest          # Unittest

import log               # Module under test


class TestLogger(unittest.TestCase):

    def test_creating_same(self):
        self.assertIs(log.Logger('hello'), log.Logger('hello'))

    def test_creating_different(self):
        self.assertIsNot(log.Logger('hello'), log.Logger('goodbye'))

    @unittest.skip("Haven't implemented this yet.")
    def test_creating_default(self):
        self.assertIsNot(log.Logger(), log.Logger())


log.py
import datetime
import sys
from threading import Lock

DEFAULT_DATE_FORMAT = "%Y-%m-%d %H:%M:%S%z"
DEFAULT_FORMAT = "[{datetime}] {level}: {message}"
DEFAULT_LEVEL = "INFO"
NEWLINE = '\n'

LEVELS = {"NOTSET":   00,
          "DEBUG":    10,
          "INFO":     20,
          "NOTICE":   25,
          "WARNING":  30,
          "ERROR":    40,
          "CRITICAL": 50}

## HELPER FUNCTIONS


def _isiterable(obj):
    try:
        iter(obj)
    except TypeError:
        return False
    else:
        return True


def _get_list(obj):
    if _isiterable(obj):
        return list(obj)
    else:
        return [obj]


def _if_none(obj, func, sentinel=None):
    return func() if obj is sentinel else obj


## FUNCTIONS

def level(level):
    level = LEVELS.get(level, level)

    def filter_func(msg_level):
        msg_level = LEVELS.get(msg_level, msg_level)
        if msg_level >= level:
            return True
        else:
            return False
    return filter_func


## Output
class Output(object):
    def __init__(self, output, **kwargs):

        # Set self.output
        if output in ['stdin', 'stdout', 'stderr']:  # Catch these seperately
            self.output = getattr(sys, output)
        elif isinstance(output, basestring):            # Gave filename
            self.output = open(output, kwargs.get("filemode", 'a'))
        else:                                        # Gave some filetype
            self.output = output

        # Set self.filters
        filter_func = kwargs.get("filter", lambda tag: True)
        self.filters = _get_list(filter_func)

        filter_funcs = kwargs.get("filters", list())
        self.filters.extend(filter_funcs)

        # Set self.level
        self.level = kwargs.get("level", level(DEFAULT_LEVEL))

    def check_valid(self, message):
        """A check to see if `message` should be written by this Output."""
        true_tags = list()
        for tag in message.tags:
            for filter_func in self.filters:
                if not filter_func(tag):
                    break
            else:
                true_tags.append(True)

        if self.level(message.level):
            return any(true_tags)

    def write(self, message):
        """Writes `message` to `self.output`"""
        self.output.write(message.line())
        self.output.write(NEWLINE)
        try:
            self.output.flush()
        except:
            pass


class Message(object):
    def __init__(self, message, level=DEFAULT_LEVEL,
                 tags=None, args=None, kwargs=None):
        self.tags = _get_list(_if_none(tags, list))
        self.raw_message = message
        args = _if_none(args, list)
        kwargs = _if_none(kwargs, dict)
        self.message = message.format(*args, **kwargs)
        self.level = level
        self.datetime = datetime.datetime.today()


class Logger(object):

    instances = {}

    def __new__(cls, name="root", *args, **kwargs):
        if name in cls.instances:
            return cls.instances[name]
        else:
            new = object.__new__(cls, *args, **kwargs)
            new.name = name
            cls.instances[name] = new
            return new

    def __init__(self, name="root", **kwargs):
        if hasattr(self, 'outputs'):
            self.outputs.extend(_get_list(kwargs.get("outputs", [])))
        else:
            self.outputs = _get_list(kwargs.get("outputs", []))

        self.lock = Lock()

    def add_outputs(self, *outputs):
        self.outputs.extend(outputs)

    def add_output(self, output):
        self.outputs.append(output)

    def _log(self, message, level, tags, args, kwargs):
        message = Message(message, tags, level, args, kwargs)
        for output in self.outputs:
            if output.check_valid(message):
                output.write(message)

    def log(self, message, level=DEFAULT_LEVEL, tags=None, *args, **kwargs):
        return self._log(message, level, tags, args, kwargs)

    def debug(self, message, tags=None, *args, **kwargs):
        return self._log(message, "DEBUG", tags, args, kwargs)

    def info(self, message, tags=None, *args, **kwargs):
        return self._log(message, "INFO", tags, args, kwargs)

    def notice(self, message, tags=None, *args, **kwargs):
        return self._log(message, "NOTICE", tags, args, kwargs)

    def warning(self, message, tags=None, *args, **kwargs):
        return self._log(message, "WARNING", tags, args, kwargs)

    def error(self, message, tags=None, *args, **kwargs):
        return self._log(message, "ERROR", tags, args, kwargs)

    def critical(self, message, tags=None, *args, **kwargs):
        return self._log(message, "CRITICAL", tags, args, kwargs)


samples/example_use.py
import log

logger = log.Logger()
logger.add_output("stdout")
logger.add_output("stderr", level=log.level("ERROR"))
logger.add_output("blue_logs.log", filter=lambda tag: tag == "blue")

logger.add_tags("green")

logger.log("Some small debug information about green.", level=log.DEBUG)
logger.log("Some small debug information about red and green.", level=log.DEBUG, tags=["red"])
logger.log("A blue error!", level=log.ERROR, tags='blue')

logger.clear_tags()

logger.log("This has no tags, and is on the default (INFO) level.")
logger.log("This has the blue tag, and is on the default level.", tags="blue")

logger.add_tags("green", "blue")

logger2 = logger.spawn()

logger2.clear_tags()

logger2.tags()
# ["green", "blue"]

logger2.clear_tags(force=True)
logger2.tags()
# []

logger.tags()
# ["green", "blue"]
logger.remove_tags("green")
logger.tags()
# ["blue"]

logger2 is logger
# False
logger.name
# "default"
logger2.name
# "default.default"

logger.spawn("sublogger").name
# default.sublogger

logger3 = log.Logger()
logger3 is logger
# False

logger4 = log.Logger("logger_name")
logger5 = log.Logger("logger_name")
logger4 is logger5
# True




log/__init__.py
import datetime
import sys
from threading import Lock

from . import five

DEFAULT_DATE_FORMAT = "%Y-%m-%d %H:%M:%S%z"
DEFAULT_FORMAT = "[{datetime}] {level}: {message}"
DEFAULT_LEVEL = "INFO"
NEWLINE = '\n'

LEVELS = {"NOTSET":   00,
          "DEBUG":    10,
          "INFO":     20,
          "NOTICE":   25,
          "WARNING":  30,
          "ERROR":    40,
          "CRITICAL": 50}

## HELPER FUNCTIONS


def _isiterable(obj):
    try:
        iter(obj)
    except TypeError:
        return False
    else:
        return True


def _get_list(obj):
    if _isiterable(obj):
        return list(obj)
    else:
        return [obj]


def _if_none(obj, func, sentinel=None):
    return func() if obj is sentinel else obj


## FUNCTIONS

def level(level):
    level = LEVELS.get(level, level)

    def filter_func(msg_level):
        msg_level = LEVELS.get(msg_level, msg_level)
        if msg_level >= level:
            return True
        else:
            return False
    return filter_func


## Output
class Output(object):
    def __init__(self, output, **kwargs):

        # Set self.output
        if output in ['stdin', 'stdout', 'stderr']:  # Catch these seperately
            self.output = getattr(sys, output)
        elif isinstance(output, five.str):           # Gave filename
            self.output = open(output, kwargs.get("filemode", 'a'))
        else:                                        # Gave some filetype
            self.output = output

        # Set self.filters
        filtr = kwargs.get("filter")
        self.filters = [] if filtr is None else [filtr]

        filter_funcs = kwargs.get("filters", list())
        self.filters.extend(filter_funcs)

        # If a message has no tags, should we accept it?
        self.accept_empty = kwargs.get("accept_empty", True)

        # Set self.level
        self.level = kwargs.get("level", level(DEFAULT_LEVEL))  # fix this

        self.format = kwargs.get('format', DEFAULT_FORMAT)
        self.date_fmt = kwargs.get('date_fmt', DEFAULT_DATE_FORMAT)
        if 'formatter' in kwargs:
            self.formatter = kwargs['formatter']

    def check_valid(self, message):
        """A check to see if `message` should be written by this Output."""
        if (not message.tags) and not self.filters:
            return True

        for filtr in self.filters:
            if not filtr(message):
                return False

        return True

    def formatter(self, message):
        time_string = message.datetime.strftime(self.date_fmt)
        args = message.args()
        args.update(datetime=time_string)
        return self.format.format(**args)

    def write(self, message):
        """Writes `message` to `self.output`"""
        self.output.write(self.formatter(message))
        self.output.write(NEWLINE)
        try:
            self.output.flush()
        except:
            pass


class Message(object):
    def __init__(self, message, level=DEFAULT_LEVEL,
                 tags=None, args=None, kwargs=None):
        self.tags = _get_list(_if_none(tags, list))
        self.raw_message = message
        args = _if_none(args, list)
        kwargs = _if_none(kwargs, dict)
        self.message = message.format(*args, **kwargs)
        self.level = level
        self.datetime = datetime.datetime.today()

    def args(self):
        return {'tags': self.tags,
                'message': self.message,
                'level': self.level,
                'datetime': self.datetime}

    def __str__(self):
        return "<{clname} {message!r}>".format(
            clname=self.__class__.__name__, message=self.message)

    __repr__ = __str__


class Logger(object):

    instances = {}

    def __new__(cls, name='root', *args, **kwargs):
        if name in cls.instances:
            return cls.instances[name]

        else:
            new = object.__new__(cls, *args, **kwargs)
            new.name = name
            cls.instances[name] = new
            return new

    def __init__(self, name="root", **kwargs):
        if hasattr(self, 'outputs'):
            self.outputs.extend(_get_list(kwargs.get("outputs", [])))
        else:
            self.outputs = _get_list(kwargs.get("outputs", []))

        if kwargs.get('output') is not None:
            self.outputs.append(kwargs['output'])

        self.lock = Lock()

    def add_outputs(self, *outputs):
        self.outputs.extend(outputs)

    def add_output(self, output):
        output = output if isinstance(output, Output) else Output(output)
        self.outputs.append(output)

    def _log(self, message, level, tags, args, kwargs):
        message = Message(message, level, tags, args, kwargs)
        for output in self.outputs:
            if output.check_valid(message):
                output.write(message)

    def log(self, message, level=DEFAULT_LEVEL, tags=None, *args, **kwargs):
        return self._log(message, level, tags, args, kwargs)

    def debug(self, message, tags=None, *args, **kwargs):
        return self._log(message, "DEBUG", tags, args, kwargs)

    def info(self, message, tags=None, *args, **kwargs):
        return self._log(message, "INFO", tags, args, kwargs)

    def notice(self, message, tags=None, *args, **kwargs):
        return self._log(message, "NOTICE", tags, args, kwargs)

    def warning(self, message, tags=None, *args, **kwargs):
        return self._log(message, "WARNING", tags, args, kwargs)

    def error(self, message, tags=None, *args, **kwargs):
        return self._log(message, "ERROR", tags, args, kwargs)

    def critical(self, message, tags=None, *args, **kwargs):
        return self._log(message, "CRITICAL", tags, args, kwargs)
